homeassistant/components/mikrotik/__init__.py

```diff
@@ -1 +1,220 @@
+            vol.Optional(CONF_LOGIN_METHOD): vol.Any(MTK_LOGIN_PLAIN, MTK_LOGIN_TOKEN),
+            vol.Optional(CONF_PORT): cv.port,
+            vol.Optional(CONF_SSL, default=False): cv.boolean,
+            vol.Optional(CONF_ENCODING, default=DEFAULT_ENCODING): cv.string,
+            vol.Optional(CONF_TRACK_DEVICES, default=True): cv.boolean,
+            vol.Optional(CONF_ARP_PING, default=False): cv.boolean,
+        }
+    )
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.All(cv.ensure_list, [MIKROTIK_SCHEMA])}, extra=vol.ALLOW_EXTRA
+)
+
+
+def setup(hass, config):
+    """Set up the Mikrotik component."""
+    hass.data[DOMAIN] = {HOSTS: {}}
+
+    for device in config[DOMAIN]:
+        host = device[CONF_HOST]
+        use_ssl = device.get(CONF_SSL)
+        user = device.get(CONF_USERNAME)
+        password = device.get(CONF_PASSWORD, "")
+        login = device.get(CONF_LOGIN_METHOD)
+        encoding = device.get(CONF_ENCODING)
+        track_devices = device.get(CONF_TRACK_DEVICES)
+
+        if CONF_PORT in device:
+            port = device.get(CONF_PORT)
+        else:
+            if use_ssl:
+                port = MTK_DEFAULT_API_SSL_PORT
+            else:
+                port = MTK_DEFAULT_API_PORT
+
+        if login == MTK_LOGIN_PLAIN:
+            login_method = (login_plain,)
+        elif login == MTK_LOGIN_TOKEN:
+            login_method = (login_token,)
+        else:
+            login_method = (login_plain, login_token)
+
+        try:
+            api = MikrotikClient(
+                host, use_ssl, port, user, password, login_method, encoding
+            )
+            api.connect_to_device()
+            hass.data[DOMAIN][HOSTS][host] = {"config": device, "api": api}
+        except (
+            librouteros.exceptions.TrapError,
+            librouteros.exceptions.MultiTrapError,
+            librouteros.exceptions.ConnectionError,
+        ) as api_error:
+            _LOGGER.error("Mikrotik %s error %s", host, api_error)
+            continue
+
+        if track_devices:
+            hass.data[DOMAIN][HOSTS][host][DEVICE_TRACKER] = True
+            load_platform(hass, DEVICE_TRACKER, DOMAIN, None, config)
+
+    if not hass.data[DOMAIN][HOSTS]:
+        return False
+    return True
+
+
+class MikrotikClient:
+    """Handle all communication with the Mikrotik API."""
+
+    def __init__(self, host, use_ssl, port, user, password, login_method, encoding):
+        """Initialize the Mikrotik Client."""
+        self._host = host
+        self._use_ssl = use_ssl
+        self._port = port
+        self._user = user
+        self._password = password
+        self._login_method = login_method
+        self._encoding = encoding
+        self._host_name = ""
+        self._info = None
+        self._client = None
+        self._connecting = False
+        self._connected = False
+
+    def connect_to_device(self):
+        """Connect to Mikrotik device."""
+        if self._connecting:
```

Comment:
When could this happen? We don't seem to use the same host api more than from one place at a time.

Reply:
That makes sense. Removing _connecting